sdk/rust/oak/src/handle.rs

```diff
@@ -0,0 +1,224 @@
+//!
+//! Applications will usually not interact with the types in this module directly, as the
+//! `HandleVisit` trait is automatically derived for all protobuf types compiled with `oak_utils`,
+//! and extracting and injecting handles is taken care of by
+//! [`oak::io::Receiver`](../io/struct.Receiver.html) and
+//! [`oak::io::Sender`](../io/struct.Sender.html).
+include!(concat!(env!("OUT_DIR"), "/oak.handle.rs"));
+
+use crate::OakError;
+
+/// Raw handle identifier.
+///
+/// This is only meant to be used with the [HandleVisit](trait.HandleVisit.html) trait, for
+/// anything else [Handle](../struct.Handle.html) should be preferred.
+// Note to maintainers: users may not add oak_abi as an explicit dependency, so we re-export this
+// type here for the custom derive to use.
+pub type Handle = oak_abi::Handle;
+
+/// Visit all handles present in a type.
+///
+/// The most common types that contains handles are
+/// [`oak::io::Receiver`](../io/struct.Receiver.html) and
+/// [`oak::io::Sender`](../io/struct.Sender.html).
+///
+/// This trait can be automatically derived:
+/// ```
+/// use oak::handle::HandleVisit;
+///
+/// #[derive(HandleVisit)]
+/// struct Thing {
+///     receiver: oak::io::Receiver<String>,
+///     sender: oak::io::Sender<String>,
+/// }
+/// ```
+///
+/// Alternatively, you can implement the trait manually. This is required if you want the trait to
+/// visit handles that are directly contained in a type.
+///
+/// ```
+/// use oak::handle::{Handle, HandleVisit};
+///
+/// struct Thing {
+///     handle: Handle,
+/// }
+///
+/// impl HandleVisit for Thing {
+///     fn visit<F: FnMut(&mut Handle)>(&mut self, mut visitor: F) -> F {
+///         visitor(&mut self.handle);
+///         visitor
+///     }
+/// }
+/// ```
+pub trait HandleVisit {
+    /// Invokes the provided closure on every handle contained in `self`.
+    fn visit<F: FnMut(&mut Handle)>(&mut self, visitor: F) -> F;
+}
+
+// A default implementation of the HandleVisit trait that does nothing
+macro_rules! handle_visit_blanket_impl {
+    ($($t:ty),+) => {
+        $(
+            impl HandleVisit for $t {
+                fn visit<F: FnMut(&mut Handle)>(&mut self, visitor: F) -> F {
+                    visitor
+                }
+            }
+        )+
+    };
+}
+
+// Provide an implementation for all scalar types in Prost.
+// See: https://github.com/danburkert/prost#scalar-values
+handle_visit_blanket_impl!((), f64, f32, i32, i64, u32, u64, bool, String, Vec<u8>);
+
+// Provide an implementation for oak_abi types that an implementation cannot be derived for.
+// These do not contains handles, so a blanket impl is sufficient.
+handle_visit_blanket_impl!(
+    oak_abi::proto::oak::encap::GrpcResponse,
+    oak_abi::proto::oak::encap::GrpcRequest,
+    oak_abi::proto::oak::log::LogMessage,
+    oak_abi::proto::oak::application::ConfigMap
+);
+
+/// Return all handles in `T`.
+///
+/// Also marks all handles in `T` invalid.
```

Comment:
"marks" how?  Maybe reword this to something like: "Extracts all handle values in `T`, replacing all of them with `oak_abi::INVALID_VALUE` (or somesuch)?